packages/core/src/Nodes/FunctionNode.ts

```diff
@@ -26,20 +27,72 @@ export class FunctionNode extends Node {
     );
   }
 }
-export class FunctionNode2 extends FunctionNode {
+
+export class FunctionDesc extends NodeDescription {
   constructor(props: {
-    description: NodeDescription;
-    graph: Graph;
-    inputs?: Socket[];
-    outputs?: Socket[];
-    exec: () => void;
+    name: string;
+    label?: string;
+    category?: NodeCategory;
+    aliases?: string[];
+    in?: { [name: string]: string }[] | string[];
+    out: { [name: string]: string }[] | string;
+    exec: (...args: any[]) => any;
   }) {
+    const alpha = 'abcdefghijklmnop';
+    const inSocketKeys: string[] = [];
+    const inSocketSpecs: { [name: string]: string } = {};
+
+    const in2 = props.in ?? [];
+    in2.forEach((inElement, index) => {
+      if (typeof inElement === 'string') {
+        inSocketSpecs[alpha[index]] = inElement;
+        inSocketKeys.push(alpha[index]);
+      } else {
+        const inElementKeys = Object.keys(inElement);
+        Assert.mustEqual(inElementKeys.length, 1);
+        inSocketSpecs[inElementKeys[0]] = inElement[inElementKeys[0]];
+        inSocketKeys.push(inElementKeys[0]);
+      }
+    });
+
+    const outSocketSpecs: { [name: string]: string } = {};
+    const outSocketKeys: string[] = [];
+
+    const out2 =
+      typeof props.out === 'string' ? [{ result: props.out }] : props.out;
+    out2.forEach((outElement) => {
+      const outElementKeys = Object.keys(outElement);
+      Assert.mustEqual(outElementKeys.length, 1);
+      outSocketSpecs[outElementKeys[0]] = outElement[outElementKeys[0]];
+      outSocketKeys.push(outElementKeys[0]);
+    });
+
     super(
-      props.description,
-      props.graph,
-      props.inputs,
-      props.outputs,
-      props.exec
+      props.name,
```

Comment:
Is it possible to not have to inherit from a root node? I think it would be better to follow what we discussed in the issue using interfaces and separating the state from the node definition?